From 70b566b97e0565ebfa94a8e043749c1cb58e1f69 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:26:37 -0400
Subject: [tools] perf: Do not accept parse_tag_value() overflow

Message-id: <1408462094-14194-229-git-send-email-jolsa@redhat.com>
Patchwork-id: 88192
O-Subject: [PATCH RHEL7.1 BZ1131394 228/325] perf tools: Do not accept parse_tag_value() overflow
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 56921becdd1eb0720603fc2e6e4c7f518196d917
Author: Adrian Hunter <adrian.hunter@intel.com>
Date: Tue Oct 22 10:34:17 2013 +0300

description
===========
parse_tag_value() accepts an "unsigned long" and multiplies it according
to a tag character.  Do not accept the value if the multiplication
overflows.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/util.c b/tools/perf/util/util.c
index c25e57b..28a0a89 100644
--- a/tools/perf/util/util.c
+++ b/tools/perf/util/util.c
@@ -386,6 +386,8 @@ unsigned long parse_tag_value(const char *str, struct parse_tag *tags)
    if (s != endptr)
     break;
 
+   if (value > ULONG_MAX / i->mult)
+    break;
    value *= i->mult;
    return value;
   }
-- 
1.7.1